xaitk_saliency/utils/bin/det_sal_on_img_dir.py

```diff
@@ -0,0 +1,123 @@
+    CONFIG_FILE - Configuration file for the DetectImageObjects, PerturbImage,
+        and GenerateDetectorProposalSaliency implementations to use.
+
+    \f
+    :param img_dir: Directory containing the images.
+    :param output_dir: Directory to write the saliency maps to.
+    :param config_file: Config file specifying the ``DetectImageObjects``,
+        ``PerturbImage``, and ``GenerateDetectorProposalSaliency``
+        implementations to use.
+    :param overlay_image: Overlay saliency maps on images. RGB images are
+        converted to grayscaleDefault is to output write out saliency maps by
+        themselves.
+    :param verbose: Display progress messages. Default is false.
+    """
+
+    if generate_config_file:
+
+        config = {}
+
+        config["DetectImageObjects"] = make_default_config(DetectImageObjects.get_impls())
+        config["PerturbImage"] = make_default_config(PerturbImage.get_impls())
+        config["GenerateDetectorProposalSaliency"] = make_default_config(GenerateDetectorProposalSaliency.get_impls())
+
+        json.dump(config, generate_config_file, indent=4)
+
+        exit()
+
+    files = os.listdir(img_dir)
+
+    img_files = []
+    imgs = []
+
+    for file_name in files:
+        try:
+            img = np.asarray(Image.open(os.path.join(img_dir, file_name)))
+            imgs.append(img)
+            img_files.append(file_name)
+        except IOError:
+            print(f"File {file_name} is not a valid image file.")
+
+    config = json.load(config_file)
+
+    blackbox_detector = from_config_dict(config["DetectImageObjects"], DetectImageObjects.get_impls())
+    img_perturber = from_config_dict(config["PerturbImage"], PerturbImage.get_impls())
+    sal_generator = from_config_dict(
+        config["GenerateDetectorProposalSaliency"],
+        GenerateDetectorProposalSaliency.get_impls()
+    )
+
+    if verbose:
+        logging.basicConfig(level=logging.INFO)
+
+    sal_maps_list = gen_detector_sal(imgs, blackbox_detector, img_perturber, sal_generator, verbose=verbose)
+
+    for img_file, ref_img, img_sal_maps in zip(img_files, imgs, sal_maps_list):
+
+        dot_ind = img_file.rfind('.')
+
+        if dot_ind == -1:
+            img_name = img_file
+        else:
+            img_name = img_file[:dot_ind]
+
+        sub_dir = os.path.join(output_dir, img_name)
+
+        os.makedirs(sub_dir, exist_ok=True)
+
+        for det_i, sal_map in enumerate(img_sal_maps):
+            fig = plt.figure()
+            plt.axis('off')
+            if overlay_image:
+                gray_img = np.asarray(Image.fromarray(ref_img).convert("L"))
+                plt.imshow(gray_img, alpha=0.7, cmap='gray')
+                plt.imshow(sal_map, cmap='jet', alpha=0.3)
+                plt.colorbar()
+            else:
+                plt.imshow(sal_map, cmap='jet')
+                plt.colorbar()
+            plt.savefig(os.path.join(sub_dir, f"det_{det_i}.jpeg"), bbox_inches='tight')
```

Comment:
I just noticed running this locally that we are not fixing the colorbar legend scale, e.g. its outputting the min-max range per-detection-sal-map, which changes almost every time. My first instinct is that we should fix the range of the color bar across all outputs to the documented output standard of [-1, 1].

@brianhhu What are your thoughts on this?

Should we maybe get this in first, do a little session of "let's try using this" and then generate derivative issues/tasks from that, instead of continuing to delay the initial introduction of the tool?

Reply:
Regarding the colorbar, my experience with RISE (at least the unbiased form, which will be true of D-RISE) is that saliency values tend to get squashed near the upper-end. So normalizing to [-1, 1] might result in pretty low-contrast, reddish saliency.

Regarding the PR itself, I think it makes sense to try to get this is in so we can do some further integration testing. We will probably have to make edits and fixes anyways as our use case evolves and we get more direction from the JAIC. Thanks!